Gems/Profiler/Code/Source/CpuProfiler.h

```diff
@@ -47,36 +54,178 @@ namespace Profiler
+    //! The CpuProfiler is able to request the cached time regions from the CpuTimingLocalStorage.
+    class CpuTimingLocalStorage
+        : public AZStd::intrusive_refcount<AZStd::atomic_uint>
     {
+        friend class CpuProfiler;
+
     public:
-        using ThreadTimeRegionMap = AZStd::unordered_map<AZStd::string, AZStd::vector<CachedTimeRegion>>;
-        using TimeRegionMap = AZStd::unordered_map<AZStd::thread_id, ThreadTimeRegionMap>;
+        AZ_CLASS_ALLOCATOR(CpuTimingLocalStorage, AZ::SystemAllocator, 0);
+
+        CpuTimingLocalStorage();
+        ~CpuTimingLocalStorage();
+
+    private:
+        // Maximum stack size
+        static constexpr uint32_t TimeRegionStackSize = 2048u;
+
+        // Adds a region to the stack, gets called each time a region begins
+        void RegionStackPushBack(CachedTimeRegion& timeRegion);
+
+        // Pops a region from the stack, gets called each time a region ends
+        void RegionStackPopBack();
+
+        // Add a new cached time region. If the stack is empty, flush all entries to the cached map
+        void AddCachedRegion(const CachedTimeRegion& timeRegionCached);
+
+        // Tries to flush the map to the passed parameter, only if the thread's mutex is unlocked
+        void TryFlushCachedMap(ThreadTimeRegionMap& cachedRegionMap);
+
+        // Clears m_cachedTimeRegions and resets m_cachedDataLimitReached flag.
+        void ResetCachedData();
+
+        AZStd::thread_id m_executingThreadId;
+        // Keeps track of the current thread's stack depth
+        uint32_t m_stackLevel = 0u;
+
+        // Cached region map, will be flushed to the system's map when the system requests it
+        ThreadTimeRegionMap m_cachedTimeRegionMap;
+
+        // Use fixed vectors to avoid re-allocating new elements
+        // Keeps track of the regions that added and removed using the macro
+        AZStd::fixed_vector<CachedTimeRegion, TimeRegionStackSize> m_timeRegionStack;
+
+        // Keeps track of regions that completed (i.e regions that was pushed and popped from the stack)
+        // Intermediate storage point for the CachedTimeRegions, when the stack is empty, all entries will be
+        // copied to the map.
+        AZStd::fixed_vector<CachedTimeRegion, TimeRegionStackSize> m_cachedTimeRegions;
+        AZStd::mutex m_cachedTimeRegionMutex;
+
+        // Dirty flag which is set when the CpuProfiler's enabled state is set from false to true
+        AZStd::atomic_bool m_clearContainers = false;
+
+        // When the thread is terminated, it will flag itself for deletion
+        AZStd::atomic_bool m_deleteFlag = false;
+
+        // Keep track of the regions that have hit the size limit so we don't have to lock to check
+        AZStd::map<AZStd::string, bool> m_hitSizeLimitMap;
+
+        // Keeps track of the first time cached data limit was reached.
+        bool m_cachedDataLimitReached = false;
```

Comment:
Nitpick: consider sorting the member variables from largest to smallest to avoid some needless padding, although you may want to place the 2 atomics further apart to avoid contention on the same cache line.